Source/pyboy/pyboy.py

```diff
@@ -124,50 +207,213 @@ def stop(self, save=True):
                     filter(itemgetter(0), zip(self.mb.cpu.hitRate, range(0x200), names)), reverse=True):
                 print("%3x %16s %s" % (n, name, hits))
 
+        if self.record_input:
+            with open(self.record_input_file, 'wb') as f:
+                recorded_data = io.StringIO() # json.dump writes 'str' not bytes...
+                json.dump(self.recorded_input, recorded_data)
+                f.write(zlib.compress(recorded_data.getvalue().encode('ascii')))
+
+
     ###################################################################
     # Scripts and bot methods
```

Comment:
I'm a little concerned at the heavy coupling this implies between `pyboy.PyBoy` and the window class(es), and that the specification for these methods is here but the implementation elsewhere. Of course, we need a way to do this, and we don't want to make an API for accessing all the ugly innards of the `window` classes, so maybe it would be better to have a `botsupport.lcd` class or something. 

(Also, as you know, I think that rendering/screen buffer should be entirely separate (in `lcd.py`) from the game window(s), but that discussion is a bit much for this already large commit...)

Reply:
We touch the subject on Discord as well. It's a balance. At the moment I think this will do. But we can review it again in the future.

Reply:
About the rendering. There might be some repeated code some places. But then again, your scanline window and my SDL2 window renders in two very different ways.